lightning/src/onion_message/packet.rs

```diff
@@ -92,25 +92,70 @@ impl LengthReadable for Packet {
 /// Onion message payloads contain "control" TLVs and "data" TLVs. Control TLVs are used to route
 /// the onion message from hop to hop and for path verification, whereas data TLVs contain the onion
 /// message content itself, such as an invoice request.
-pub(super) enum Payload {
+pub(super) enum Payload<T: OnionMessageContents> {
 	/// This payload is for an intermediate hop.
 	Forward(ForwardControlTlvs),
 	/// This payload is for the final hop.
 	Receive {
 		control_tlvs: ReceiveControlTlvs,
 		reply_path: Option<BlindedRoute>,
-		// Coming soon:
-		// message: Message,
+		message: Message<T>,
 	}
 }
 
-// Coming soon:
-// enum Message {
-// 	InvoiceRequest(InvoiceRequest),
-// 	Invoice(Invoice),
-//	InvoiceError(InvoiceError),
-//	CustomMessage<T>,
-// }
+/// Trait to be implemented by custom onion messages.
+pub trait CustomMessageReadable {
+	/// Decodes a custom message given the message type. If the given message type is known to the
+	/// implementation and the message could be decoded, must return `Ok(Some(message))`. If the
+	/// message type is unknown to the implementation, must return `Ok(None)`. If a decoding error
+	/// occur, must return `Err(DecodeError::X)` where `X` details the encountered error.
+	fn read<R: io::Read>(message_type: u64, buffer: &mut R) -> Result<Option<Self>, DecodeError> where Self: Sized;
+}
+
+#[derive(Debug)]
+/// The contents of an onion message. In the context of offers, this would be the invoice, invoice
+/// request, or invoice error.
+pub enum Message<T> where T: OnionMessageContents {
+	// Coming soon:
+	// Invoice,
+	// InvoiceRequest,
+	// InvoiceError,
+	/// A custom onion message specified by the user.
+	Custom(T),
+}
+
+impl<T> Message<T> where T: OnionMessageContents {
+	/// Returns the type that was used to decode the message payload.
+	pub fn tlv_type(&self) -> u64 {
+		match self {
+			&Message::Custom(ref msg) => msg.tlv_type(),
+		}
+	}
+}
+
+impl<T: OnionMessageContents> Writeable for Message<T> {
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
+		match self {
+			Message::Custom(msg) => Ok(msg.write(w)?),
+		}
+	}
+}
+
+/// Defines a type identifier for encoding an onion message's contents.
+pub trait OnionMessageContents: Writeable + CustomMessageReadable {
+	/// Returns the type identifying the message payload.
+	fn tlv_type(&self) -> u64;
+}
+
+impl CustomMessageReadable for () {
+	fn read<R: io::Read>(_message_type: u64, _buffer: &mut R) -> Result<Option<Self>, DecodeError> where Self: Sized {
+		Ok(None)
+	}
+}
+
+impl OnionMessageContents for () {
+	fn tlv_type(&self) -> u64 { u64::max_value() }
+}
```

Comment:
We probably don't want to allow writing `()` as the onion message. Could we either separate the traits or use an enum with no variants? Then this could be `unreachable`.